{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "plink2_path ='/ Users/cameronwalker/Documents/CosnetiX/plink2'\n",
    "class PlinkPipeline:\n",
    "    def __init__(self, input_prefix, output_prefix):\n",
    "        self.plink2_path = r\"C:\\Users\\greys\\OneDrive\\Downloads\\plink2\\plink2.exe\"\n",
    "        self.input_prefix = input_prefix\n",
    "        self.output_prefix = output_prefix\n",
    "\n",
    "    def run_command(self, cmd):\n",
    "        subprocess.run(cmd, check=True)\n",
    "\n",
    "    def quality_control(self):\n",
    "        cmd = [self.plink2_path, \"--bfile\", self.input_prefix, \"--maf\", \"0.01\",\n",
    "               \"--geno\", \"0.05\", \"--hwe\", \"1e-6\", \"--make-bed\", \"--out\", \"temp\"]\n",
    "        self.run_command(cmd)\n",
    "\n",
    "    def filter_samples_and_variants(self):\n",
    "        cmd = [self.plink2_path, \"--bfile\", \"temp\", \"--mind\",\n",
    "               \"0.1\", \"--geno\", \"0.05\", \"--make-bed\", \"--out\", \"temp2\"]\n",
    "        self.run_command(cmd)\n",
    "\n",
    "    def format_output(self):\n",
    "        cmd = [self.plink2_path, \"--bfile\", \"temp2\",\n",
    "               \"--recode\", \"vcf\", \"--out\", self.output_prefix]\n",
    "        self.run_command(cmd)\n",
    "\n",
    "    def run_pipeline(self):\n",
    "        self.quality_control()\n",
    "        self.filter_samples_and_variants()\n",
    "        self.format_output()\n",
    "        \n",
    "\n",
    "# Adjust this to the actual path of your plink2 executable\n",
    "plink2_path = r\"C:\\path\\to\\plink2.exe\"\n",
    "input_prefix = 'output'  # The prefix of the input BED, BIM, and FAM files\n",
    "output_prefix = 'final_output'  # The prefix of the output VCF file\n",
    "txt_input_file = 'sample.txt'  # The path to the .txt file you want to convert\n",
    "\n",
    "pipeline =\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "def quality_control(input_file, output_prefix):\n",
    "    # Filter based on MAF, missing genotype rate and HWE p-value\n",
    "    qc1_output = output_prefix + \"_qc1\"\n",
    "    qc1_command = f\"plink --bfile {input_file} --maf 0.01 --geno 0.05 --hwe 1e-6 --make-bed --out {qc1_output}\"\n",
    "    subprocess.call(qc1_command, shell=True)\n",
    "    \n",
    "    # Filter out samples with missing genotype rate > 0.1\n",
    "    qc2_output = output_prefix + \"_qc2\"\n",
    "    qc2_command = f\"plink --bfile {qc1_output} --mind 0.1 --make-bed --out {qc2_output}\"\n",
    "    subprocess.call(qc2_command, shell=True)\n",
    "\n",
    "    # Filter out variants with missing genotype rate > 0.05\n",
    "    qc3_output = output_prefix + \"_qc3\"\n",
    "    qc3_command = f\"plink --bfile {qc2_output} --geno 0.05 --make-bed --out {qc3_output}\"\n",
    "    subprocess.call(qc3_command, shell=True)\n",
    "    \n",
    "    # Convert cleaned data to VCF format\n",
    "    vcf_output = output_prefix + \"_cleaned\"\n",
    "    vcf_command = f\"plink --bfile {qc3_output} --recode vcf --out {vcf_output}\"\n",
    "    subprocess.call(vcf_command, shell=True)\n",
    "    \n",
    "    # Remove intermediate files\n",
    "    remove_files = [\"bed\", \"bim\", \"fam\", \"log\"]\n",
    "    for ext in remove_files:\n",
    "        for i in range(1, 4):\n",
    "            file = f\"{output_prefix}_qc{i}.{ext}\"\n",
    "            if os.path.isfile(file):\n",
    "                os.remove(file)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    input_file = \"your_input_file\" # Replace with your input file name (in PLINK binary format)\n",
    "    output_prefix = \"your_output_prefix\" # Replace with your desired output prefix\n",
    "    quality_control(input_file, output_prefix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\greys\\AppData\\Local\\Temp\\ipykernel_27548\\2209136692.py:12: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  self.data = pd.read_csv(self.input_file, sep='\\t', header=None, names=[\n",
      "c:\\Users\\greys\\.envs\\cosnetix-webapp\\Lib\\site-packages\\pandas\\core\\series.py:893: RuntimeWarning: invalid value encountered in cast\n",
      "  return np.asarray(self._values, dtype)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "\n",
    "class BedConverter:\n",
    "    def __init__(self, input_file, output_prefix):\n",
    "        self.input_file = input_file\n",
    "        self.output_prefix = output_prefix\n",
    "\n",
    "    def load_data(self):\n",
    "        # Load data from the input file\n",
    "        self.data = pd.read_csv(self.input_file, sep='\\t', header=None, names=[\n",
    "                                'rsid', 'chromosome', 'position', 'genotype'])\n",
    "\n",
    "    def convert_genotypes(self):\n",
    "        # Convert the genotypes to numeric values\n",
    "        genotype_map = {'AA': 0, 'CC': 0, 'GG': 0, 'TT': 0,\n",
    "                        'AC': 1, 'AG': 1, 'AT': 1, 'CG': 1, 'CT': 1, 'GT': 1}\n",
    "        self.data['genotype_numeric'] = self.data['genotype'].map(genotype_map)\n",
    "\n",
    "    def create_bed_file(self):\n",
    "        # Create the binary BED file\n",
    "        num_snps = len(self.data)\n",
    "        num_individuals = 1\n",
    "        bed_file = f\"{self.output_prefix}.bed\"\n",
    "\n",
    "        # Write the BED file header\n",
    "        with open(bed_file, 'wb') as bed:\n",
    "            bed.write(b'\\x6C\\x1B\\x01')\n",
    "\n",
    "            # Write the genotype data\n",
    "            for i in range(0, num_snps, 4):\n",
    "                snp_data = self.data['genotype_numeric'].iloc[i:i+4]\n",
    "                packed_byte = np.packbits(\n",
    "                    np.array(snp_data * 2, dtype=np.uint8))\n",
    "                bed.write(packed_byte)\n",
    "\n",
    "    def convert(self):\n",
    "        self.load_data()\n",
    "        self.convert_genotypes()\n",
    "        self.create_bed_file()\n",
    "\n",
    "    def quality_filter(self):\n",
    "        cmd = [plink2_path, \"--bfile\", input_file_path, \"--maf\", \"0.01\",\n",
    "            \"--geno\", \"0.05\", \"--hwe\", \"1e-6\", \"--make-bed\", \"--out\", \"temp\"]\n",
    "        subprocess.run(cmd, check=True)\n",
    "\n",
    "input_file = r'C:\\Users\\greys\\projects\\cosnetix_pipeline\\external\\genome\\genomicSample.txt'\n",
    "output_prefix = 'yeet_bitch'\n",
    "\n",
    "converter = BedConverter(input_file, output_prefix)\n",
    "converter.convert()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/bin/sh: /: is a directory\n",
      "/bin/sh: /: is a directory\n",
      "/bin/sh: /: is a directory\n",
      "/bin/sh: /: is a directory\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "class QualityControl:\n",
    "    def __init__(self, input_file, output_prefix):\n",
    "        self.input_file = input_file\n",
    "        self.output_prefix = output_prefix\n",
    "        self.plink2_path ='/ Users/cameronwalker/Documents/CosnetiX/plink2'\n",
    "    def run(self):\n",
    "        # Filter based on MAF, missing genotype rate and HWE p-value\n",
    "        qc1_output = self.output_prefix + \"_qc1\"\n",
    "        qc1_command = f\"{self.plink2_path} --bfile {self.input_file} --maf 0.01 --geno 0.05 --hwe 1e-6 --make-bed --out {qc1_output}\"\n",
    "        subprocess.call(qc1_command, shell=True)\n",
    "        \n",
    "        # Filter out samples with missing genotype rate > 0.1\n",
    "        qc2_output = self.output_prefix + \"_qc2\"\n",
    "        qc2_command = f\"{self.plink2_path} --bfile {qc1_output} --mind 0.1 --make-bed --out {qc2_output}\"\n",
    "        subprocess.call(qc2_command, shell=True)\n",
    "\n",
    "        # Filter out variants with missing genotype rate > 0.05\n",
    "        qc3_output = self.output_prefix + \"_qc3\"\n",
    "        qc3_command = f\"{self.plink2_path} --bfile {qc2_output} --geno 0.05 --make-bed --out {qc3_output}\"\n",
    "        subprocess.call(qc3_command, shell=True)\n",
    "        \n",
    "        # Convert cleaned data to VCF format\n",
    "        vcf_output = self.output_prefix + \"_cleaned\"\n",
    "        vcf_command = f\"{self.plink2_path} --bfile {qc3_output} --recode vcf --out {vcf_output}\"\n",
    "        subprocess.call(vcf_command, shell=True)\n",
    "        \n",
    "        # Remove intermediate files\n",
    "        remove_files = [\"bed\", \"bim\", \"fam\", \"log\"]\n",
    "        for ext in remove_files:\n",
    "            for i in range(1, 4):\n",
    "                file = f\"{self.output_prefix}_qc{i}.{ext}\"\n",
    "                if os.path.isfile(file):\n",
    "                    os.remove(file)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    input_file = \"/Users/cameronwalker/Documents/GitHub/cosnetix_pipeline/yeet_bitch\" # Replace <input_file> with your input file name (in PLINK binary format)\n",
    "    output_prefix = \"yeet_bitches\" # Replace yeet_bitches with your desired output prefix\n",
    "    qc = QualityControl(input_file, output_prefix)\n",
    "    qc.run()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'C:\\\\Users\\\\greys\\\\projects\\\\cosnetix_pipeline\\\\external\\\\genome\\\\genomicSample.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 61\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m--make-bed completed in: \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39moutput_prefix\u001b[39m=}\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m     57\u001b[0m converter \u001b[39m=\u001b[39m PlinkConverter(\n\u001b[1;32m     58\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mC:\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mUsers\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgreys\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mprojects\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mcosnetix_pipeline\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mexternal\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgenome\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgenomicSample.txt\u001b[39m\u001b[39m\"\u001b[39m, \n\u001b[1;32m     59\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mgenomic-demo\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     60\u001b[0m     \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mC:\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mUsers\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgreys\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mprojects\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mcosnetix_pipeline\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mpipeline\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgwas\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mdata\u001b[39m\u001b[39m\\\u001b[39m\u001b[39minterim\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m---> 61\u001b[0m converter\u001b[39m.\u001b[39;49mconvert_to_ped_map()\n\u001b[1;32m     62\u001b[0m converter\u001b[39m.\u001b[39mrun_make_bed()\n",
      "Cell \u001b[0;32mIn[4], line 18\u001b[0m, in \u001b[0;36mPlinkConverter.convert_to_ped_map\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     16\u001b[0m sample_data \u001b[39m=\u001b[39m {}\n\u001b[1;32m     17\u001b[0m genotype_data \u001b[39m=\u001b[39m {}\n\u001b[0;32m---> 18\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49minput_file, \u001b[39m\"\u001b[39;49m\u001b[39mr\u001b[39;49m\u001b[39m\"\u001b[39;49m) \u001b[39mas\u001b[39;00m input_file:\n\u001b[1;32m     19\u001b[0m     header \u001b[39m=\u001b[39m input_file\u001b[39m.\u001b[39mreadline()\u001b[39m.\u001b[39mstrip()\u001b[39m.\u001b[39msplit(\u001b[39m\"\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     20\u001b[0m     \u001b[39mfor\u001b[39;00m line \u001b[39min\u001b[39;00m input_file:\n",
      "File \u001b[0;32m/opt/homebrew/anaconda3/envs/cosnetix/lib/python3.11/site-packages/IPython/core/interactiveshell.py:284\u001b[0m, in \u001b[0;36m_modified_open\u001b[0;34m(file, *args, **kwargs)\u001b[0m\n\u001b[1;32m    277\u001b[0m \u001b[39mif\u001b[39;00m file \u001b[39min\u001b[39;00m {\u001b[39m0\u001b[39m, \u001b[39m1\u001b[39m, \u001b[39m2\u001b[39m}:\n\u001b[1;32m    278\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    279\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mIPython won\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt let you open fd=\u001b[39m\u001b[39m{\u001b[39;00mfile\u001b[39m}\u001b[39;00m\u001b[39m by default \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    280\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mas it is likely to crash IPython. If you know what you are doing, \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    281\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39myou can use builtins\u001b[39m\u001b[39m'\u001b[39m\u001b[39m open.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    282\u001b[0m     )\n\u001b[0;32m--> 284\u001b[0m \u001b[39mreturn\u001b[39;00m io_open(file, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'C:\\\\Users\\\\greys\\\\projects\\\\cosnetix_pipeline\\\\external\\\\genome\\\\genomicSample.txt'"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "from os.path import join \n",
    "\n",
    "class PlinkConverter:\n",
    "    def __init__(self, input_file, output_prefix, output_path):\n",
    "        self.input_file = input_file\n",
    "        self.output_prefix = output_prefix\n",
    "        self.output_path = output_path\n",
    "        self.plink_path = r\"C:\\Users\\greys\\OneDrive\\Downloads\\plink2\\plink2.exe\"\n",
    "        \n",
    "    def convert_to_ped_map(self):\n",
    "        # Read in the input file and convert to PED and MAP formats\n",
    "        # Replace this with your own code to parse the input file and create PED and MAP files\n",
    "\n",
    "        # Example code to read in sample and genotype data from a tab-separated file\n",
    "        sample_data = {}\n",
    "        genotype_data = {}\n",
    "        with open(self.input_file, \"r\") as input_file:\n",
    "            header = input_file.readline().strip().split(\"\\t\")\n",
    "            for line in input_file:\n",
    "                fields = line.strip().split(\"\\t\")\n",
    "                sample_id = fields[0]\n",
    "                phenotype = fields[1]\n",
    "                sample_data[sample_id] = phenotype\n",
    "                genotypes = fields[2:]\n",
    "                genotype_data[sample_id] = genotypes\n",
    "\n",
    "        ped_file = join(self.output_path, f\"{self.output_prefix}.ped\")\n",
    "        map_file = join(self.output_path, f\"{self.output_prefix}.map\")\n",
    "        # Create the PED file\n",
    "        with open(ped_file, \"w\") as ped_out:\n",
    "            # Write out the sample IDs and phenotype values\n",
    "            for sample_id, phenotype in sample_data.items():\n",
    "                ped_out.write(\n",
    "                    f\"{sample_id}\\t{sample_id}\\t0\\t0\\t0\\t{phenotype}\\t\")\n",
    "                # Write out the genotype data\n",
    "                for genotype in genotype_data[sample_id]:\n",
    "                    ped_out.write(f\"{genotype}\\t\")\n",
    "                ped_out.write(\"\\n\")\n",
    "        # Create the MAP file\n",
    "        with open(map_file, \"w\") as map_out:\n",
    "            for i, snp_id in enumerate(header[2:], start=1):\n",
    "                map_out.write(f\"0\\t{snp_id}\\t0\\t{i}\\n\")\n",
    "                \n",
    "        print(f'PED and MAP files created in: {self.output_path=}')\n",
    "        \n",
    "    def run_make_bed(self):\n",
    "        # Call the plink2 executable to convert the PED and MAP files to binary PLINK format\n",
    "        subprocess.run([self.plink_path, \"ped\",\n",
    "                   f\"{self.output_prefix}.ped\", f\"{self.output_prefix}.map\", \"--make-bed\", f\"--out={self.output_prefix}.bed\"])\n",
    "\n",
    "        # subprocess.run([self.plink_path, \"--make-bed\",\n",
    "        #                f\"--file={self.output_prefix}\", f\"--out={self.output_prefix}\"])\n",
    "        \n",
    "        print(f'--make-bed completed in: {self.output_prefix=}')\n",
    "        \n",
    "converter = PlinkConverter(\n",
    "    r\"C:\\Users\\greys\\projects\\cosnetix_pipeline\\external\\genome\\genomicSample.txt\", \n",
    "    r\"genomic-demo\",\n",
    "    r\"C:\\Users\\greys\\projects\\cosnetix_pipeline\\pipeline\\gwas\\data\\interim\")\n",
    "converter.convert_to_ped_map()\n",
    "converter.run_make_bed()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 52\u001b[0m\n\u001b[0;32m     50\u001b[0m input_file \u001b[39m=\u001b[39m \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mC:\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mUsers\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgreys\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mprojects\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mcosnetix_pipeline\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mexternal\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgenome\u001b[39m\u001b[39m\\\u001b[39m\u001b[39mgenomicSample.txt\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m     51\u001b[0m output_prefix \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mgenomic_sample\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m---> 52\u001b[0m convert_sample_to_ped_map(input_file, output_prefix)\n\u001b[0;32m     55\u001b[0m \u001b[39m# Example usage\u001b[39;00m\n\u001b[0;32m     56\u001b[0m ped_file \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mgenomic_sample.ped\u001b[39m\u001b[39m\"\u001b[39m\n",
      "Cell \u001b[1;32mIn[7], line 21\u001b[0m, in \u001b[0;36mconvert_sample_to_ped_map\u001b[1;34m(input_file, output_prefix)\u001b[0m\n\u001b[0;32m     17\u001b[0m         \u001b[39mmap\u001b[39m\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mchromosome\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mrsid\u001b[39m}\u001b[39;00m\u001b[39m 0 \u001b[39m\u001b[39m{\u001b[39;00mposition\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     19\u001b[0m         \u001b[39m# Write .ped file\u001b[39;00m\n\u001b[0;32m     20\u001b[0m         \u001b[39m# Assuming a generic \"FAM001\" family ID, \"IND001\" individual ID, \"0\" paternal and maternal IDs, \"1\" for sex (male), and \"-9\" for missing phenotype\u001b[39;00m\n\u001b[1;32m---> 21\u001b[0m         ped\u001b[39m.\u001b[39mwrite(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mFAM001 IND001 0 0 1 -9 \u001b[39m\u001b[39m{\u001b[39;00mgenotype[\u001b[39m0\u001b[39m]\u001b[39m}\u001b[39;00m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00mgenotype[\u001b[39m1\u001b[39;49m]\u001b[39m}\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     23\u001b[0m \u001b[39mprint\u001b[39m(\n\u001b[0;32m     24\u001b[0m     \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mConversion complete. Output files are \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mped_file\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m and \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mmap_file\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[1;31mIndexError\u001b[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pandas_plink import write_plink\n",
    "\n",
    "\n",
    "def convert_ped_map_to_bed(input_ped, input_map, output_prefix):\n",
    "    # Read .ped file\n",
    "    ped_data = pd.read_csv(input_ped, sep=\" \", header=None)\n",
    "    fam_data = pd.DataFrame(ped_data.iloc[:, :6])\n",
    "    fam_data.columns = [\"fid\", \"iid\", \"father\", \"mother\", \"sex\", \"pheno\"]\n",
    "\n",
    "    # Read .map file\n",
    "    map_data = pd.read_csv(input_map, sep=\"\\t\", header=None, names=[\n",
    "                           \"chrom\", \"snp\", \"cm\", \"pos\"])\n",
    "    bim_data = map_data[[\"chrom\", \"snp\", \"cm\", \"pos\", \"pos\", \"pos\"]].copy()\n",
    "    bim_data.columns = [\"chrom\", \"snp\", \"cm\", \"pos\", \"a1\", \"a2\"]\n",
    "\n",
    "    # Extract genotype data from .ped file\n",
    "    genotype_data = ped_data.iloc[:, 6:].values.reshape(-1, len(map_data), 2)\n",
    "\n",
    "    # Write .bed, .bim, and .fam files\n",
    "    write_plink(output_prefix, (fam_data, bim_data,\n",
    "                genotype_data), verbose=True)\n",
    "\n",
    "\n",
    "# Example usage\n",
    "input_ped = \"path/to/your/input.ped\"\n",
    "input_map = \"path/to/your/input.map\"\n",
    "output_prefix = \"path/to/your/output_prefix\"\n",
    "\n",
    "convert_ped_map_to_bed(input_ped, input_map, output_prefix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "self.plink_path='C:\\\\Users\\\\greys\\\\OneDrive\\\\Downloads\\\\plink2\\\\plink2.exe'\n",
      "conversion to ped and map\n",
      "Conversion complete. Output files are 'C:\\Users\\greys\\projects\\cosnetix_pipeline\\pipeline\\gwas\\data\\interim\\_genome_sample.ped' and 'C:\\Users\\greys\\projects\\cosnetix_pipeline\\pipeline\\gwas\\data\\interim\\_genome_sample.map'.\n",
      "Error: plink2 failed to preprocess the genome data.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "from pipeline.data.helpers import read_yaml\n",
    "\n",
    "class GenotypingPreprocessor:\n",
    "    def __init__(self, data_contract_path:str, input_file:str):\n",
    "        \n",
    "        self.data_contract = read_yaml(data_contract_path)\n",
    "        \n",
    "        self.input_file = input_file\n",
    "        self.output_prefix = self.data_contract['Genome']['output']\n",
    "        # self.output_path = self.data_contract['Genome']['collections']['interim']\n",
    "        self.output_path = r'C:\\Users\\greys\\projects\\cosnetix_pipeline\\pipeline\\gwas\\data\\interim'\n",
    "        self.ped_file = os.path.join(self.output_path,f\"{self.output_prefix}.ped\")\n",
    "        self.map_file = os.path.join(self.output_path,f\"{self.output_prefix}.map\")\n",
    "        self.bed_file = os.path.join(self.output_path,f\"{self.output_prefix}.bed\")\n",
    "        self.bim_file = os.path.join(self.output_path,f\"{self.output_prefix}.bim\")\n",
    "        self.fam_file = os.path.join(self.output_path,f\"{self.output_prefix}.fam\")\n",
    "        self.output_file = os.path.join(self.output_path,f\"{self.output_prefix}.bed\")\n",
    "        self.cleaned_prefix = f\"{self.output_prefix}_cleaned\"\n",
    "        self.vcf_file = f\"{self.cleaned_prefix}.vcf\"\n",
    "        self.plink_path = os.getenv(\"PLINK_PATH\")\n",
    "        # self.plink_path = 'plink'\n",
    "        print(f'{self.plink_path=}')\n",
    "    def _check_input_file(self):\n",
    "        if not os.path.isfile(self.input_file):\n",
    "            print(f\"Error: Input file '{self.input_file}' not found.\")\n",
    "            return False\n",
    "        return True\n",
    "\n",
    "    def _check_plink2(self):\n",
    "        try:\n",
    "            subprocess.run([self.plink_path, \"--version\"], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "            return True\n",
    "        except FileNotFoundError:\n",
    "            print(\"Error: plink2 not found. Please install plink2 and ensure it's in your PATH.\")\n",
    "            return False\n",
    "\n",
    "    def convert_to_ped_map(self):\n",
    "        print('conversion to ped and map')\n",
    "        if not self._check_input_file():\n",
    "            return\n",
    "\n",
    "        with open(self.input_file, 'r') as infile, open(self.ped_file, 'w') as ped, open(self.map_file, 'w') as map:\n",
    "            for line in infile:\n",
    "                # Ignore comment lines\n",
    "                if line.startswith(\"#\"):\n",
    "                    continue\n",
    "\n",
    "                rsid, chromosome, position, genotype = line.strip().split('\\t')\n",
    "                # Write .map file\n",
    "                map.write(f\"{chromosome} {rsid} 0 {position}\\n\")\n",
    "\n",
    "                # Write .ped file\n",
    "                ped.write(f\"{rsid} {chromosome} {position} {genotype}\\n\")\n",
    "\n",
    "        print(f\"Conversion complete. Output files are '{self.ped_file}' and '{self.map_file}'.\")\n",
    "\n",
    "    def preprocess_with_plink2(self):\n",
    "        self.convert_to_ped_map()\n",
    "        if not (self._check_input_file() and self._check_plink2()):\n",
    "            return\n",
    "\n",
    "        # Convert .ped and .map files to .bed, .bim, and .fam files using plink2\n",
    "        plink2_preprocess_command = [\n",
    "            self.plink_path,\n",
    "            \"--file\", self.ped_file,\n",
    "            \"--make-bed\",\n",
    "            \"--out\", self.output_file\n",
    "        ]\n",
    "        try:\n",
    "            subprocess.run(plink2_preprocess_command, check=True)\n",
    "            print(f\"Preprocessing complete. Output files are '{self.bed_file}', '{self.bim_file}', and '{self.fam_file}'.\")\n",
    "        except subprocess.CalledProcessError:\n",
    "            print(\"Error: plink2 failed to preprocess the genome data.\")\n",
    "            return\n",
    "\n",
    "        # Quality control and filtering\n",
    "        plink2_filter_command = [\n",
    "            self.plink_path,\n",
    "            \"--bfile\", self.output_prefix,\n",
    "            \"--maf\", \"0.01\",\n",
    "            \"--geno\", \"0.05\",\n",
    "            \"--hwe\", \"1e-6\",\n",
    "            \"--mind\", \"0.1\",\n",
    "            \"--make-bed\",\n",
    "            \"--out\", self.cleaned_prefix\n",
    "        ]\n",
    "        try:\n",
    "            subprocess.run(plink2_filter_command, check=True)\n",
    "            print(f\"Quality control and filtering complete. Output files are '{self.cleaned_prefix}.bed', '{self.cleaned_prefix}.bim', and '{self.cleaned_prefix}.fam'.\")\n",
    "        except subprocess.CalledProcessError:\n",
    "            print(\"Error\")\n",
    "\n",
    "\n",
    "prc = GenotypingPreprocessor(\n",
    "    r'C:\\Users\\greys\\projects\\cosnetix_pipeline\\data-contract.yml',\n",
    "    r'C:\\Users\\greys\\projects\\cosnetix_pipeline\\external\\genome\\genomicSample.txt'\n",
    ")\n",
    "prc._check_plink2()\n",
    "prc.preprocess_with_plink2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "# Define file paths\n",
    "plink2_path = \"/Users/cameronwalker/Documents/CosnetiX/plink2\"\n",
    "input_file_path = \"/Users/cameronwalker/Downloads/genome_Bernie_Innocenti_v3_v5_Full_20230329120354.txt\"\n",
    "output_file_path = \"/Users/cameronwalker/Documents/GitHub/cosnetix_pipeline/external/gwas\"\n",
    "\n",
    "# Quality control\n",
    "cmd = [\n",
    "    plink2_path, \n",
    "    \"--txt\", input_file_path, \n",
    "    \"--maf\", \"0.01\", \n",
    "    \"--geno\", \"0.05\", \n",
    "    \"--hwe\", \"1e-6\", \n",
    "    \"--make-bed\", \n",
    "    \"--out\", \"temp\"\n",
    "    ]\n",
    "subprocess.run(cmd, check=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mRunning cells with 'Python 3.11.0 ('cosnetix')' requires ipykernel package.\n",
      "\u001b[1;31mRun the following command to install 'ipykernel' into the Python environment. \n",
      "\u001b[1;31mCommand: 'conda install -n cosnetix ipykernel --update-deps --force-reinstall'"
     ]
    }
   ],
   "source": [
    "# Filter by sample and variant\n",
    "cmd = [plink2_path, \"--bfile\", \"temp\", \"--mind\", \"0.1\", \"--geno\", \"0.05\", \"--make-bed\", \"--out\", \"temp2\"]\n",
    "subprocess.run(cmd, check=True)\n",
    "\n",
    "# Formatting\n",
    "cmd = [plink2_path, \"--bfile\", \"temp2\", \"--recode\", \"vcf\", \"--out\", output_file_path]\n",
    "subprocess.run(cmd, check=True)\n",
    "\n",
    "# Clean up intermediate files\n",
    "subprocess.run([\"rm\", \"temp*\"], check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conversion complete. Output files are 'genotyping_output.ped' and 'genotyping_output.map'.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('genotyping_output.ped', 'genotyping_output.map')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "def convert_genomic_sample_to_ped_map(input_file: str, output_prefix: str):\n",
    "    \"\"\"\n",
    "    Convert a genomicSample.txt file into .ped and .map files.\n",
    "\n",
    "    :param input_file: Path to the genomicSample.txt file\n",
    "    :param output_prefix: Prefix for the output .ped and .map files\n",
    "    \"\"\"\n",
    "    ped_file = f\"{output_prefix}.ped\"\n",
    "    map_file = f\"{output_prefix}.map\"\n",
    "\n",
    "    # Check if the input file exists\n",
    "    if not os.path.isfile(input_file):\n",
    "        print(f\"Error: Input file '{input_file}' not found.\")\n",
    "        return\n",
    "\n",
    "    with open(input_file, 'r') as infile, open(ped_file, 'w') as ped, open(map_file, 'w') as map:\n",
    "        for line in infile:\n",
    "            # Ignore comment lines\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "\n",
    "            rsid, chromosome, position, genotype = line.strip().split('\\t')\n",
    "            # Write .map file\n",
    "            map.write(f\"{chromosome} {rsid} 0 {position}\\n\")\n",
    "\n",
    "            # Write .ped file\n",
    "            ped.write(f\"{rsid} {chromosome} {position} {genotype}\\n\")\n",
    "\n",
    "    print(f\"Conversion complete. Output files are '{ped_file}' and '{map_file}'.\")\n",
    "    return ped_file, map_file\n",
    "# Example usage:\n",
    "input_file = \"/Users/cameronwalker/Documents/GitHub/cosnetix_pipeline/pipeline/gwas/genomicSample.txt\"\n",
    "output_prefix = \"genotyping_output\"\n",
    "convert_genomic_sample_to_ped_map(input_file, output_prefix)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conversion complete. Output files are 'genotyping_output.ped' and 'genotyping_output.map'.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "# Define file paths\n",
    "plink2_path = \"/Users/cameronwalker/Documents/CosnetiX/plink2\"\n",
    "\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "import os\n",
    "\n",
    "class GenotypingPreprocessor:\n",
    "    def __init__(self, input_file: str, output_prefix: str):\n",
    "        self.input_file = input_file\n",
    "        self.output_prefix = output_prefix\n",
    "        self.ped_file = f\"{output_prefix}.ped\"\n",
    "        self.map_file = f\"{output_prefix}.map\"\n",
    "        self.plink = \"/Users/cameronwalker/Documents/CosnetiX/plink2\"\n",
    "    def _check_input_file(self):\n",
    "        if not os.path.isfile(self.input_file):\n",
    "            print(f\"Error: Input file '{self.input_file}' not found.\")\n",
    "            return False\n",
    "        return True\n",
    "\n",
    "    def convert_to_ped_map(self):\n",
    "        if not self._check_input_file():\n",
    "            return\n",
    "\n",
    "        with open(self.input_file, 'r') as infile, open(self.ped_file, 'w') as ped, open(self.map_file, 'w') as map:\n",
    "            for line in infile:\n",
    "                # Ignore comment lines\n",
    "                if line.startswith(\"#\"):\n",
    "                    continue\n",
    "\n",
    "                rsid, chromosome, position, genotype = line.strip().split('\\t')\n",
    "                # Write .map file\n",
    "                map.write(f\"{chromosome} {rsid} 0 {position}\\n\")\n",
    "\n",
    "                # Write .ped file\n",
    "                ped.write(f\"{rsid} {chromosome} {position} {genotype}\\n\")\n",
    "\n",
    "        print(f\"Conversion complete. Output files are '{self.ped_file}' and '{self.map_file}'.\")\n",
    "\n",
    "# Example usage:\n",
    "input_file_path = \"/Users/cameronwalker/Downloads/genome_Bernie_Innocenti_v3_v5_Full_20230329120354.txt\"\n",
    "output_file_path = \"/Users/cameronwalker/Documents/GitHub/cosnetix_pipeline/external/gwas\"\n",
    "\n",
    "preprocessor = GenotypingPreprocessor(input_file, output_prefix)\n",
    "preprocessor.convert_to_ped_map()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('cosnetix')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "f96009708aa938ac1d0887de64ea2f8a25ad56a27052f9bd1644b09aed241525"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
